qcodes/dataset/dond/do_nd_utils.py

```diff
@@ -44,12 +44,24 @@ def _register_parameters(
     meas: Measurement,
     param_meas: Sequence[ParamMeasT],
     setpoints: Sequence[ParameterBase] | None = None,
-    shapes: Shapes = None,
+    shapes: Shapes | None = None,
 ) -> None:
-    for parameter in param_meas:
-        if isinstance(parameter, ParameterBase):
-            meas.register_parameter(parameter, setpoints=setpoints)
-    meas.set_shapes(shapes=shapes)
+    real_parameters = [
+        param for param in param_meas if isinstance(param, ParameterBase)
+    ]
+    parameter_names = [param.full_name for param in real_parameters]
+
+    for param in real_parameters:
+        if isinstance(param, MultiParameter):
+            parameter_names.extend(param.full_names)
+
+    for parameter in real_parameters:
+        meas.register_parameter(parameter, setpoints=setpoints)
+    if shapes is not None:
+        filtered_shapes = {
+            name: shape for name, shape in shapes.items() if name in parameter_names
+        }
+        meas.set_shapes(shapes=filtered_shapes)
```

Comment:
a suggestion to separate the logic needed for registering from the logic needed for the shape handling
```suggestion
    real_parameters = [
        param for param in param_meas if isinstance(param, ParameterBase)
    ]

    for parameter in real_parameters:
        meas.register_parameter(parameter, setpoints=setpoints)

    if shapes is not None:
        parameter_names = [param.full_name for param in real_parameters]
        for param in real_parameters:
            if isinstance(param, MultiParameter):
                parameter_names.extend(param.full_names)

        filtered_shapes = {
            name: shape for name, shape in shapes.items() if name in parameter_names
        }
        meas.set_shapes(shapes=filtered_shapes)
```